### PR TITLE
BTRFS file system for MAIN and rootfs partition

### PR DESCRIPTION
**Pull request by tiopex**
BTRFS file system for MAIN and rootfs partition for data inconsistency resilience

- umtdrd to share MAIN, BOOT and rootfs partition via handheld USB
- to access BTRFS partition in Windows via SD card,  required installation driver https://github.com/maharmstone/btrfs

Transfer via USB can be increased by using 4-bit dtb file

USB gadgets refactored, as umtprd require 3 endpoints and serial console require 2 endpoints.

F1c100 have only 4 endpoints so we can't use MTP and serial console together ( mass storage also required only 2 endpoints)

### CODE REVIEW

File: board/miyoo/boot/firstboot [hunk 1]
```diff
@@ -493,76 +493,19 @@ if (dialog --clear --ok-label NO --cancel-label YES --stdout --title "   License
 	safe_poweroff_func
 fi
 mount -o remount,ro "${MOUNTDIR}" | tee -a ${LOG}
-
-#grep all partitions current size
```
Comment (Apaczer) [inline board/miyoo/boot/firstboot]: This was mostly for over resizing but was nice to have a sanity check and to don't run unncessary fatresize when no unnalocated space left. In addition it was nice to have info in log.txt about what size have been allocated.

File: board/miyoo/boot/firstboot [hunk 2]
```diff
@@ -493,76 +493,19 @@ if (dialog --clear --ok-label NO --cancel-label YES --stdout --title "   License
...
-			SD_MAX_AVAIL_SIZE_IN_KiB=123731968
-		fi
-	fi
-	SD_MAX_AVAIL_SIZE_IN_GiB=$(echo "scale=1; ${SD_MAX_AVAIL_SIZE_IN_KiB} / 1024 / 1024" | bc) #only to inform user about end point for resize
-	SD_MAX_AVAIL_SIZE_IN_GB=$(echo "scale=1; ${SD_MAX_AVAIL_SIZE_IN_KiB} * 1024 / 1000 / 1000 / 1000" | bc)
-	P4_SIZE_IN_GB=$(echo "scale=1; ${P4_SIZE} * 512 / 1000 / 1000 / 1000" | bc)
-	# roundup GB value to 0,1GB for integer calculations:
-	SD_MAX_AVAIL_SIZE_IN_GBx10=$(echo "$SD_MAX_AVAIL_SIZE_IN_GB" | tr -d ".")
-	P4_SIZE_IN_GBx10=$(echo "$P4_SIZE_IN_GB" | tr -d ".")
-	#Sanity checks if we don't down- or over-resize partition:
```
Comment (Apaczer) [inline board/miyoo/boot/firstboot]: Here sanity test, we might just leave `test "$SD_MAX_UNALL_SIZE_IN_MiB" -le 15` part

File: board/miyoo/boot/firstboot [hunk 3]
```diff
@@ -577,40 +520,22 @@ mv "${MOUNTDIR}/firstboot" "${MOUNTDIR}/firstboot.done" | tee -a ${LOG}
...
-			fi
-		fi
-	done
-fi
-echo "Renaming ${HOME}/retroarch directory to ${HOME}/.retroarch" | tee -a ${LOG}
-mv "${HOME}/retroarch" "${HOME}/.retroarch" | tee -a ${LOG}
 
-echo
+echo "fix permsissions"  | tee -a ${LOG}
+chmod -R +x /mnt/  | tee -a ${LOG}
```
Comment (Apaczer) [inline board/miyoo/boot/firstboot:573]: can't we fix it from buildroot post-build script? I see you changed exe permision on config files.

File: board/miyoo/boot/firstboot [hunk 4]
```diff
@@ -493,76 +493,19 @@ if (dialog --clear --ok-label NO --cancel-label YES --stdout --title "   License
...
-		fsck.fat -y /dev/mmcblk0p4 | tee -a ${LOG}
-		echo -n -e \\xEB\\x58\\x90 > /dev/mmcblk0p4
-		sync
-		mount -t vfat -o rw,sync,utf8 /dev/mmcblk0p4 ${HOME} | tee -a ${LOG}
-	else
-		echo "MAIN fat partition resizing was aborted!" | tee -a ${LOG}
-		sleep 2
-	fi
+	echo "Resizing the MAIN partition." | tee -a ${LOG}
+	echo "This can take few seconds." | tee -a ${LOG}
```
Comment (Apaczer) [inline board/miyoo/boot/firstboot]: "a few seconds" typo

File: board/miyoo/main/apps/fsck/fsck.sh [hunk 5]
```diff
@@ -13,18 +13,6 @@ echo -e "\e[32mChecking FAT32 (BOOT) partition...\e[0m"
 fsck.vfat -a /dev/mmcblk0p1
```
Comment (Apaczer) [inline board/miyoo/main/apps/fsck/fsck.sh]: mv to main, remove script

File: board/miyoo/main/options.cfg [hunk 6]
```diff
@@ -1,5 +1,4 @@
 MODULES_CUSTOM=0 
-FAT_CHECK=1
```
Comment (Apaczer) [inline board/miyoo/main/options.cfg:2]: btrfs check(8)? How we can check data integrity?

File: board/miyoo/main/apps/usb-peripheral/usb-peripheral.sh
Comment (Apaczer) [inline board/miyoo/main/apps/usb-peripheral/usb-peripheral.sh:1]: why keep usb-peripheral & usb-mtd separate? Case use of usb-peripheral other than mass storage? If so you want to switch to peripheral automaticly while launching umtprd

File: board/miyoo/rootfs/etc/main [hunk 7]
```diff
@@ -76,20 +76,6 @@ elif test -r "${HOME}/firstboot.completed"; then
 	rm "${HOME}/firstboot.completed"
 fi
 
-if !(grep -q FAT_CHECK\=\0 "${HOME}/options.cfg"); then
```
Comment (Apaczer) [inline board/miyoo/rootfs/etc/main:79]: leave it for mmcblk0p1 ?

**Review by Apaczer — CHANGES_REQUESTED** [review]
So far good with BTRFS, had issue while unmounting bare SD inserted on Win 8.1. On the bright side I can even see rootfs partition tree with umtprd!

Couldn't remap gamepad with joymap, but it was with musl build which I haven't tested before.

File: board/miyoo/main/options.cfg (hunk 6 above)
Comment (tiopex) [reply board/miyoo/main/options.cfg:2]: this is the beauty of btrfs, you dont have to manually check, as it is done automatically

File: board/miyoo/boot/firstboot (hunk 3 above)
Comment (tiopex) [reply board/miyoo/boot/firstboot:573]: I tried, but not working

**Comment by nfriedly** [comment]
Thanks! I will try to test this out tonight.

File: board/miyoo/boot/firstboot [hunk 8]
```diff
@@ -548,21 +540,14 @@ if (dialog --clear --stdout --ok-label YES --cancel-label NO --title "   AUTO-RE
...
-		mount -t vfat -o rw,sync,utf8 /dev/mmcblk0p4 ${HOME} | tee -a ${LOG}
-	else
-		echo "MAIN fat partition resizing was aborted!" | tee -a ${LOG}
-		sleep 2
-	fi
+	echo "Resizing the MAIN partition to ${SD_MAX_AVAIL_SIZE_IN_GiB}GB." | tee -a ${LOG}
+	echo "This can take a few seconds." | tee -a ${LOG}
+	umount /dev/mmcblk0p4 | tee -a ${LOG}
+	parted -a optimal /dev/mmcblk0 --script rm 4  | tee -a ${LOG}
+	parted -a optimal /dev/mmcblk0 --script mkpart primary 669MB 100%  | tee -a ${LOG}
```
Comment (Apaczer) [inline board/miyoo/boot/firstboot]: Use non arbitrary value for start point, smth like:
`START_POINT=(${P1_SIZE} + ${P2_SIZE} + ${P3_SIZE} + ${P4_SIZE}) + $BOOT_SIZE` but in MB

**Comment by Apaczer** [comment]
I can't at times poweroff handheld with /bin/poweroff (it may be due to gmenu2x requesting sync() before shutdown), but I guess no use for it now. Joymap works on uClibc.

**Review by Apaczer — APPROVED** [review]

**Pull request merged**